[Skia] Implement Font::applyTransforms with HarfBuzz

carlosgcampos opened the pull request:
#### 4de8bbf1b6e56273042a80fcabfb7df7f95fef4c
<pre>
[Skia] Implement Font::applyTransforms with HarfBuzz
<a href="https://bugs.webkit.org/show_bug.cgi?id=297088">https://bugs.webkit.org/show_bug.cgi?id=297088</a>

Reviewed by NOBODY (OOPS!).

Revert 299354@main and implement Font::applyTransforms instead, trying
to skip the harfbuzz shapping when possible.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4de8bbf1b6e56273042a80fcabfb7df7f95fef4c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121333 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41030 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31689 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127772 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73414 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123209 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41732 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49609 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92166 "16 failures") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61326 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124285 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33326 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108725 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72842 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32343 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26858 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71352 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/113463 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102819 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27034 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130607 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/119853 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48261 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36694 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100763 "Found 101 new test failures: editing/deleting/5144139-2.html editing/selection/14971.html fast/borders/rtl-border-04.html fast/borders/rtl-border-05.html fast/css/font-face-locally-installed.html fast/css/vertical-text-overflow-ellipsis-text-align-center-mixed.html fast/css/vertical-text-overflow-ellipsis-text-align-justify-mixed.html fast/css/vertical-text-overflow-ellipsis-text-align-left-mixed.html fast/css/vertical-text-overflow-ellipsis-text-align-right-mixed.html fast/dom/Document/CaretRangeFromPoint/caretRangeFromPoint-in-zoom-and-scroll.html ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48629 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104919 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100668 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46080 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24140 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44955 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48119 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53832 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150015 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47591 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38146 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50937 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49273 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3498ad696878817ba826caba353d228f3ed6ed47. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50802

aperezdc commented on Source/WebCore/platform/graphics/skia/FontSkia.cpp:
```diff
@@ -177,4 +178,97 @@ bool Font::platformSupportsCodePoint(char32_t character, std::optional<char32_t>
+    Vector<hb_feature_t> featuresWithKerning;
+    if (!enableKerning) {
+        featuresWithKerning.reserveInitialCapacity(featuresSize + 1);
+        static hb_feature_t kernFeature { HB_TAG('k', 'e', 'r', 'n'), 0, 0, static_cast<unsigned>(-1) };
```
This could be `const`, or even `constexpr`:

```suggestion
        static constexpr hb_feature_t kernFeature { HB_TAG('k', 'e', 'r', 'n'), 0, 0, static_cast<unsigned>(-1) };
```

carlosgcampos committed:
[Skia] Implement Font::applyTransforms with HarfBuzz
https://bugs.webkit.org/show_bug.cgi?id=297088

Reviewed by NOBODY (OOPS!).

Revert 299354@main and implement Font::applyTransforms instead, trying
to skip the harfbuzz shapping when possible.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4de8bbf1b6e56273042a80fcabfb7df7f95fef4c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50802
